proxy/http.go

```diff
@@ -0,0 +1,196 @@
+package proxy
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"golang.org/x/net/http2"
+	"html/template"
+	"net/http"
+	"net/http/httputil"
+	"strconv"
+	"strings"
+
+	"github.com/estuary/data-plane-gateway/auth"
+	"github.com/estuary/flow/go/labels"
+	log "github.com/sirupsen/logrus"
+	"github.com/valyala/fasthttp"
+	"github.com/valyala/fasthttp/fasthttpadaptor"
+)
+
+func (h *ProxyHandler) proxyHttp(ctx context.Context, clientConn *tls.Conn, proxyConn *ProxyConnection, portConfig *labels.PortConfig) error {
+	defer proxyConn.Close()
+	// We generally assume that the upstream connector container wishes to speak http/1.1, unless they explicitly request to use _only_ h2.
+	var useHttp2Upstream = portConfig != nil && portConfig.Protocol == "h2"
+	var isPublicPort = portConfig != nil && portConfig.Public
+
+	var targetScheme = "http"
+	if useHttp2Upstream {
+		targetScheme = "https"
+	}
+	var proxy = httputil.ReverseProxy{
+		Transport:     proxyConn.singleConnectionTransport(useHttp2Upstream),
+		FlushInterval: 0,
+		ErrorHandler: func(w http.ResponseWriter, req *http.Request, err error) {
+			log.WithFields(log.Fields{
+				"hostname":   proxyConn.hostname,
+				"remoteAddr": clientConn.RemoteAddr().String(),
+				"shardID":    proxyConn.shardID,
+				"error":      err,
+				"URI":        req.RequestURI,
+			}).Error("proxy error")
+			handleHttpError(err, w, req)
+		},
+
+		Director: func(req *http.Request) {
+			// TODO: verify that request Host matches the value of the connection SNI
```

Comment:
Not sure if this todo is something you wanted to take care of here?

Reply:
Backstory is that I had at one time thought to dynamically re-establish upstream connections between the gateway and the connector, so that we didn't need to close the client connection whenever a connector restarts. Doing that would require this validation in order to prevent request smuggling. Anyway, I decided to just close the client connection whenever there's an error with the upstream, and missed removing this TODO.